interfaces/builtin/multipass_support.go

```diff
@@ -114,15 +125,54 @@ lchown
 lchownat
 `
 
+type multipassSupportInterface struct {
+	commonInterface
+}
+
+func (iface *multipassSupportInterface) AppArmorPermanentPlug(spec *apparmor.Specification, plug *snap.PlugInfo) error {
+	// Generate rules to allow snap-update-ns to mount apparmor config
+	emit := spec.AddUpdateNSf
+	for _, target := range apparmorConfigDirs {
+		base := plug.Snap.Base
+		if base == "" {
+			base = "core"
+		}
+		source := fmt.Sprintf("/snap/%s/*", base) + target
+		emit("  # access to %s\n", target)
+		emit("  mount options=(bind) %s/ -> %s/,\n", source, target)
+		emit("  umount %s/,\n\n", target)
+	}
+	return nil
+}
+
+func (iface *multipassSupportInterface) MountPermanentPlug(spec *mount.Specification, plug *snap.PlugInfo) error {
+	// preserve apparmor config from the core snap
+	for _, dir := range apparmorConfigDirs {
+		base := plug.Snap.Base
+		if base == "" {
+			base = "core"
+		}
+		basePath, _ := filepath.EvalSymlinks(fmt.Sprintf("/snap/%s/current", base))
```

Comment:
The `/snap` path is not a constant, there are two possible locations. It'd best to use `dirs.SnapMountDir` instead. Keep in mind that the apparmor profiles should use `/snap` as that variability only exists on the host mount namespace.

Having said that, a longer comment: I was really worried about this part as it means the mount profile encodes the full path, with revision, of a base snap. This means that each time the base snap changes, this should change as well.

Fortunately, snap-confine has logic for discarding mount namespaces and managing the transition when that is required. In addition snap-update-ns can update the mount namespace, hopefully correctly, in a case when the base snap changes revision but cannot be discarded.

I think this is actually correct and safe. We should add a test that refreshes the base snap (just reinstall current core revision) to ensure nothing explodes. I will write a separate longer comment about that, as doing the exact right thing is tricky (we can easily end up in one of the easy cases that do not exercise quite what I mean here).

Reply:
Thanks - I changed the hard-coded `/snap` to `dirs.SnapMountDir` in 3bbc2477bd - we can try address the other test later.